File: android/android/src/de/benibela/videlibri/RenewList.java
package de.benibela.videlibri;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;

import java.util.ArrayList;

import de.benibela.videlibri.jni.Bridge;

public class RenewList extends BookListActivity {
    public Button button;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setTitle(tr(R.string.renew_title_start));
        selectedBooks = new ArrayList<>();
        updateViewFilters();


        button = (Button) findViewById(R.id.buttonbelowlist);
        //if (button == null) return;
        button.setVisibility(View.VISIBLE);
        button.setEnabled(false);
        button.setText(tr(R.string.booklist_noselection));
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                VideLibriApp.renewBooks(selectedBooks.toArray(new Bridge.Book[0]));
                finish();
            }
        });
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateViewFilters();
        displayBookCache();
    }

    @Override
    public void onBookCacheAvailable() {
        super.onBookCacheAvailable();
        updateRenewButton();
    }

    private int truecount;

    private void updateViewFilters(){
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);
        setOption(BookOverviewAdapter.DisplayEnum.NoDetails, sp.getBoolean("noLendBookDetails", false));
        setOption(BookOverviewAdapter.DisplayEnum.ShowRenewCount, sp.getBoolean("showRenewCount", true));
        sortingKey = sp.getString("sorting", "dueDate");
        groupingKey = sp.getString("grouping", "_dueWeek");
        ArrayList<Bridge.Book> oldSelection = selectedBooks;
        bookCache = LendingList.makePrimaryBookCache(false, true);
        truecount = bookCache.size();
        bookCache = LendingList.filterToSecondaryBookCache(bookCache, groupingKey, sortingKey, "", null);
        selectedBooks = new ArrayList<>(selectedBooks.size());
        for (Bridge.Book selbook: oldSelection)
            for (Bridge.Book book: bookCache)
                if (selbook.equalsBook(book)) {
                    selectedBooks.add(book);
                    break;
                }
    }

    @Override
    public void viewDetails(int bookpos) {
        boolean deleted = false;
        for (int i=0;i<selectedBooks.size(); i++ )
            if (selectedBooks.get(i) == bookCache.get(bookpos))  {
                selectedBooks.remove(i);
                deleted = true;
                break;
            }
        if (!deleted)
            selectedBooks.add(bookCache.get(bookpos));
        updateRenewButton();
        ((BookOverviewAdapter)((ListView) findViewById(R.id.booklistview)).getAdapter()).notifyDataSetChanged();
        if (!port_mode) super.viewDetails(bookpos);
    }


    private void updateRenewButton(){
        if (selectedBooks.size() == 0) {
            setTitle(tr(R.string.renew_title_select));
            if (button != null) {
                button.setEnabled(false);
                button.setText(tr(R.string.renew_noselection));
            }
        } else {
            setTitle(tr(R.string.renew_title_selectionDD, selectedBooks.size(), truecount));
            if (button != null) {
                button.setEnabled(true);
                button.setText(tr(R.string.renew_renewD, selectedBooks.size()));
                //button.setBackgroundDrawable(activity.getResources().getDrawable(android.R.drawable.butt)); //http://stackoverflow.com/questions/4384890/how-to-disable-an-android-button
            }
        }
    }
}

File: android/android/src/de/benibela/videlibri/RenewList.kt
package de.benibela.videlibri

import android.os.Bundle
import android.preference.PreferenceManager
import android.view.View
import android.widget.Button
import android.widget.ListView
import de.benibela.videlibri.jni.Bridge
import java.util.ArrayList

class RenewList : BookListActivity() {
    lateinit var button: Button

    private var truecount: Int = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        title = tr(R.string.renew_title_start)
        selectedBooks = ArrayList()
        updateViewFilters()


        button = findViewById(R.id.buttonbelowlist)
        button.apply {
            visibility = View.VISIBLE
            isEnabled = false
            text = tr(R.string.booklist_noselection)
            setOnClickListener {
                VideLibriApp.renewBooks(selectedBooks?.toTypedArray() ?: return@setOnClickListener)
                finish()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        updateViewFilters()
        displayBookCache()
    }

    override fun onBookCacheAvailable() {
        super.onBookCacheAvailable()
        updateRenewButton()
    }

    private fun updateViewFilters() {
        val sp = PreferenceManager.getDefaultSharedPreferences(this)
        displayOptions.readFromPreferences(sp)
        displayOptions.filterKey = ""
        val oldSelection = selectedBooks
        bookCache = makePrimaryBookCache(false, true)
        truecount = bookCache.size
        bookCache = displayOptions.let { filterToSecondaryBookCache(bookCache, it.groupingKey, it.sortingKey, "", "") }
        val newSelection = ArrayList<Bridge.Book>(selectedBooks?.size?:0)
        oldSelection?.forEach {selbook ->
            for (book in bookCache)
                if (selbook.equalsBook(book)) {
                    newSelection.add(book)
                    break
                }
        }
        selectedBooks = newSelection
    }

    override fun viewDetails(bookpos: Int) {
        val book = bookCache[bookpos]
        val deleteIndex = selectedBooks?.indexOfFirst{ it === book} ?: -1
        if (deleteIndex == -1) selectedBooks?.add(book)
        else selectedBooks?.removeAt(deleteIndex)
        updateRenewButton()
        (findViewById<ListView>(R.id.booklistview).adapter as BookOverviewAdapter).notifyDataSetChanged()
        if (!port_mode) super.viewDetails(bookpos)
    }


    private fun updateRenewButton() {
        if (selectedBooks?.size == 0) {
            title = tr(R.string.renew_title_select)
            button.isEnabled = false
            button.text = tr(R.string.renew_noselection)
        } else {
            title = tr(R.string.renew_title_selectionDD, selectedBooks?.size, truecount)
            button.isEnabled = true
            button.text = tr(R.string.renew_renewD, selectedBooks?.size)
        }
    }
}
